src/mpid/ch4/src/ch4r_rma_target_callbacks.c

```diff
@@ -1655,27 +1646,27 @@ int MPIDIG_put_data_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_
     int mpi_errno = MPI_SUCCESS;
     MPIR_Request *rreq;
     MPIDIG_put_dat_msg_t *msg_hdr = (MPIDIG_put_dat_msg_t *) am_hdr;
-    struct iovec *iov;
-    uintptr_t base;
-    int i;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_PUT_DATA_TARGET_MSG_CB);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_PUT_DATA_TARGET_MSG_CB);
     MPIR_T_PVAR_TIMER_START(RMA, rma_targetcb_put_data);
 
     rreq = (MPIR_Request *) msg_hdr->preq_ptr;
-    base = (uintptr_t) MPIDIG_REQUEST(rreq, req->preq.target_addr);
 
-    /* Adjust the target iov addresses using the base address
-     * (window base + target_disp) */
-    iov = (struct iovec *) MPIDIG_REQUEST(rreq, req->preq.dt_iov);
-    for (i = 0; i < MPIDIG_REQUEST(rreq, req->preq.n_iov); i++)
-        iov[i].iov_base = (char *) iov[i].iov_base + base;
+    /* FIXME: MPIR_Typerep_unflatten should allocate the new object */
+    MPIR_Datatype *dt = (MPIR_Datatype *) MPIR_Handle_obj_alloc(&MPIR_Datatype_mem);
```

Comment:
There is no thread contention inside an am handler, so I think can safely malloc the datatype object and attach the pointer to the request. The `MPIR_Handle_obj_alloc` will invoke a critical section inside (in the per-vci thread model).

If we do `malloc`, there is no need to `MPIR_Object_set_ref`, or reference the handle.

Reply:
I don't think there's any instance where we malloc a `MPIR_Datatype` instead of use the object allocator. I'm not sure I want to start in this PR. I do think it would be better if the typerep code did this for us, though.

Reply:
Typerep code has no context such as whether a datatype handle is needed or whether the reference counting is needed.

Reply:
Typerep is being asked to create a datatype from flattened representation. All datatypes have handles. I think it's reasonable to start from a single reference and let the caller modify the refcount, if necessary. It is essentially a datatype creation function.

Reply:
> All datatypes have handles.

That is not true. Handles are only needed for user interface. Internally all operations are on `MPIR_Datatype` structure.

I guess it is okay to use handle object in this PR and leave it for later performance tuning. Just to note here that we are incurring a critical section that is going to be in contention with all object routines including request objects in a performance critical path.

Reply:
> > All datatypes have handles.
> 
> That is not true. Handles are only needed for user interface. Internally all operations are on `MPIR_Datatype` structure.

Handles are used in most device APIs. All datatypes have handles today, and I don't really want to change that right now.

Reply:
Please use the typerep API.  That sets the refcounts correctly.  There's a handle allocated, but is not published because it's for internal use only.

Reply:
This usage is copy/paste from ch3 RMA code. I do not see where the typerep code sets refcount during unflatten. Maybe I am missing something...

Reply:
Oh, you are right.  The caller sets the refcount in typerep.  `yaksa` sets the refcount internally, which confused me a bit.